### PR TITLE
Adds a LinterMessage component

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-code-manager/issues/295

Here is a sampling of the variants:

screen shot 2019-03-04 at 9 37 22 am

### CODE REVIEW

**Comment by kumar303** [comment]
Consider this in a draft state (I guess you can't switch to draft state?). I will fix up https://github.com/mozilla/addons-code-manager/pull/302 and rebase on that first.

**Comment by willdurand** [comment]
Looks like there is an escaping issue in the message:

screen shot 2019-03-05 at 16 04 11

File: src/components/LinterMessage/index.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import { Alert } from 'react-bootstrap';
+
+import styles from './styles.module.scss';
+import { LinterMessage as LinterMessageType } from '../../linter';
+
+export type PublicProps = {
+  description: LinterMessageType['description'];
+  message: LinterMessageType['message'];
+  type: LinterMessageType['type'];
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: Question: why not passing a `LinterMessage` directly?

File: src/components/LinterMessage/index.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+const getAlertVariant = (type: PublicProps['type']) => {
+  switch (type) {
+    case 'error':
+      return 'danger';
+    case 'warning':
+      return 'warning';
+    default:
+    case 'notice':
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: `default` is positioned before so I don't think this is needed.

File: src/components/LinterMessage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      return 'secondary';
+  }
+};
+
+const urlPattern = /^https?:\/\//;
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: Why is this extracted outside the function? It's used once, so I think we can inline it.

File: src/components/LinterMessage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+const renderDescriptionPart = (part: string) => {
+  // Split a description into words to intercept and replace
+  // URLs with JSX links.
+  return part
+    .trim()
+    .split(' ')
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: Would it make sense to split on `\s` instead?

File: src/components/LinterMessage/index.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    <Alert variant={getAlertVariant(type)}>
+      <Alert.Heading>{message}</Alert.Heading>
+      <p className={styles.description}>
+        {description.reduce((lines: React.ReactNode[], text, index) => {
+          if (index > 0) {
+            lines.push(<br key={text} />);
+          }
+          lines.push(renderDescriptionPart(text));
+          return lines;
+        }, [])}
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: Why is `renderDescriptionPart()` outside the component but not this `reduce()`? It'd be easier to read if we could have a `renderDescription()` function (that could maybe be a single function, embedding the current `renderDescriptionPart()` impl).

File: src/components/LinterMessage/index.tsx (hunk 4 above)
Comment (kumar303) [reply src/components/LinterMessage/index.tsx]: Whoa, `.split(/(\s)/)` magically adds the space back! I didn't know that.

> If separator is a regular expression that contains capturing parentheses, then each time separator is matched, the results (including any undefined results) of the capturing parentheses are spliced into the output array.

File: src/components/LinterMessage/index.tsx (hunk 3 above)
Comment (kumar303) [reply src/components/LinterMessage/index.tsx]: It's an old Python habit of only compiling the regex once but who knows how JS works. I'll move it inline.

File: src/components/LinterMessage/index.tsx (hunk 5 above)
Comment (kumar303) [reply src/components/LinterMessage/index.tsx]: I moved the reducers into one function and that does read better. Thanks.

**Comment by kumar303** [comment]
> Looks like there is an escaping issue in the message

D'oh. It worked at first. Surprisingly, React escapes when a prop has nested strings (like an array or object) but not when it's just a string 🤷‍♀️

**Comment by kumar303** [comment]
OK, yeah, I'm renaming this to `src/reducers/linter` because I'll definitely need it there for https://github.com/mozilla/addons-code-manager/issues/41

**Comment by kumar303** [comment]
> I'm renaming this to src/reducers/linter

omg, I'm losing my mind. I don't know why I did that in this branch. Reverted 😛 I'll do it in the next patch.

**Comment by willdurand** [comment]
> omg, I'm losing my mind. I don't know why I did that in this branch. Reverted 😛 I'll do it in the next patch.

ok, you can also do that in a small patch. easy to review.

File: src/components/LinterMessage/index.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    expect(root.find(Alert.Heading)).toHaveText(message);
+    expect(root.find(`.${styles.description}`)).toIncludeText(description[0]);
+  });
+
+  it('handles descriptions with extra whitespace', () => {
+    const description = ['  lots    of    space     between'];
+    const root = renderMessage({ description });
+
+    expect(root.find(`.${styles.description}`)).toIncludeText(
+      description[0].trim(),
```
Comment (willdurand) [inline src/components/LinterMessage/index.spec.tsx]: If I understand correctly, it only trims each description entry. The value of `description` in this test case makes me think that, maybe, the extra spaces between each words are also trimmed, but that's not the case, right?

File: src/components/LinterMessage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+import unescape from 'lodash.unescape';
+import { Alert } from 'react-bootstrap';
+
+import styles from './styles.module.scss';
+import { LinterMessage as LinterMessageType } from '../../linter';
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: If you rename the component `LinterMessageBase`, you could avoid this alias.

File: src/components/LinterMessage/index.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    allLines.push(
+      // Intercept and replace URLs with JSX links.
+      line
+        .trim()
+        .split(/(\s+)/)
+        .reduce((allParts: React.ReactNode[], part) => {
+          if (urlPattern.test(part)) {
+            allParts.push(<Alert.Link href={part}>{part}</Alert.Link>);
```
Comment (willdurand) [inline src/components/LinterMessage/index.tsx]: Don't we need a `key` here too? 🤔

Comment (willdurand) [reply src/components/LinterMessage/index.tsx]: Will these links be internal or external? I am thinking about react-router and target=_blank.

Comment (kumar303) [reply src/components/LinterMessage/index.tsx]: They will be external, yeah

File: src/components/LinterMessage/index.spec.tsx (hunk 6 above)
Comment (kumar303) [reply src/components/LinterMessage/index.spec.tsx]: Actually, this doesn't really need trim at all

File: src/components/LinterMessage/index.tsx (hunk 8 above)
Comment (kumar303) [reply src/components/LinterMessage/index.tsx]: I will file an issue to make the links external because I don't see an obvious way to do it

**Pull request merged**